fix(supervisor): fix return value when process is already registered

thalesmg opened the pull request:
This fixes a bug where the return type for `supervisor:start_child/2` was not being respected (sort of).

If a supervisor tries to start a child who tries to register the same local name that a running process has already registered, the return value was of the form `{error, {{already_started, pid()}, child_spec()}}` rather than `{error, {already_started, pid()}}`.  

Although `term()` allows that, it can be a bit misleading.

IngelaAndin commented:
You need to also add the testcase name to the callback all/0,  that lists all test cases, for the tests to be run.

thalesmg commented:
> @thalesmg You need to also add the testcase name to the callback all/0, that lists all test cases, for the tests to be run.

Oops, I had added the test case there, but forgot to add to the commit.  :see_no_evil: 
Fixed, thanks.  :beers:

IngelaAndin commented:
Thanks for the fix it will be tested again after RC3 (release candidate for OTP-26) and Easter holidays.

IngelaAndin commented:
Now it fails like this in our test runs:

```
*** CT Error Notification 2023-04-27 01:11:04.431 ***[🔗](http://otp.ericsson.se/product/internal/test/test_results/progress_26/2023_04_26/otp_26_artemis_linux-gnu_x86_64_64-bit_s32:4_kp_a1_emu_docker_64517_26/ct_run.test_server@artemis.2023-04-27_00.27.27/test.stdlib_test.logs/run.2023-04-27_00.29.42/supervisor_suite.already_started_outside_supervisor.html#e-6436)
supervisor_SUITE:already_started_outside_supervisor failed on line 3683
Reason: {badmatch,{error,{already_started,<0.342605.0>}}}

[Full error description and stacktrace](http://otp.ericsson.se/product/internal/test/test_results/progress_26/2023_04_26/otp_26_artemis_linux-gnu_x86_64_64-bit_s32:4_kp_a1_emu_docker_64517_26/ct_run.test_server@artemis.2023-04-27_00.27.27/test.stdlib_test.logs/run.2023-04-27_00.29.42/supervisor_suite.already_started_outside_supervisor.html#end)



=== Ended at 2023-04-27 01:11:04
=== Location: [{supervisor_SUITE,already_started_outside_supervisor,[3683](http://otp.ericsson.se/product/internal/test/test_results/progress_26/2023_04_26/otp_26_artemis_linux-gnu_x86_64_64-bit_s32:4_kp_a1_emu_docker_64517_26/ct_run.test_server@artemis.2023-04-27_00.27.27/test.stdlib_test.logs/run.2023-04-27_00.29.42/supervisor_suite.src.html#3683)},
              {test_server,ts_tc,1782},
              {test_server,run_test_case_eval1,1291},
              {test_server,run_test_case_eval,1223}]
=== === Reason: no match of right hand side value 
                 {error,{already_started,<0.342605.0>}}
  in function  supervisor_SUITE:already_started_outside_supervisor/1 (supervisor_SUITE.erl, line 3683)
  in call from test_server:ts_tc/3 (test_server.erl, line 1782)
  in call from test_server:run_test_case_eval1/6 (test_server.erl, line 1291)
  in call from test_server:run_test_case_eval/9 (test_server.erl, line 1223)
```

thalesmg commented:
Strange, when I run this test alone locally it seems to work.  That line suggests that some other test is not stopping the registered `sup_test` supervisor.  :thinking: 

I'll try to reproduce and fix, thanks.

thalesmg commented:
I could not reproduce the issue locally, but I've tried to add a check that stops the test supervisor if it is still alive from another testcase run.

IngelaAndin commented:
I will test it again!

thalesmg commented:
Sure, I'll investigate it and fix as soon as possible.  :smiley_cat:

thalesmg commented:
Hi @IngelaAndin .  Unfortunately, I was not able to reproduce locally, and the logs from CI doesn't indicate why those 2 tests from `gen_server_SUITE` are failing.  :sweat: 

Locally, test 1 is skipped ("no OTP 25 available" ?) and 2 just passes:

![image](https://github.com/erlang/otp/assets/16166434/112f758d-cf84-41eb-9a76-2e9af770e106)

I tried to search the log output downloaded from CI, but it just seems to indicate that the tests fail due to a timetrap timeout while calling `compile:file` on a remote node running an older OTP?  At this [line](https://github.com/erlang/otp/blob/d01cfe15aef9c9a7548e52265d5c14cd8e6d93ba/lib/stdlib/test/gen_server_SUITE.erl#L1556).

Would you have any tips on how I could gather more logs from CI to help understand why the peer node hung?

IngelaAndin commented:
I have looked into this a little and it seems unrelated to your PR . I will make some more test to try and verify that!

thalesmg commented:
> I have looked into this a little and it seems unrelated to your PR . I will make some more test to try and verify that!

Thanks!  :beers:

IngelaAndin commented:
@thalesmg  It looks good in our nightly builds so I think we can ignore this as being an unrelated problem connected to GitHub actions run. I  am now thinking that you should rebase this on maint so that it can get into OTP-26 track and not get it first in OTP-27.

IngelaAndin commented:
Thanks for the PR :)

thalesmg commented:
Thank you for your review!  :beers:
